marlowe-dashboard-client/src/Contract/State.purs

```diff
@@ -400,10 +404,54 @@ regenerateStepCards currentSlot state =
 
     previousSteps = confirmedSteps <> pendingTimeoutSteps
 
-    namedActions = extractNamedActions currentSlot (state ^. _executionState)
+    executionState = state ^. _executionState
+
+    userParties = state ^. _userParties
+
+    participants = state ^. _participants
+
+    namedActions =
+      expandAndGroupByRole
+        userParties
+        (Map.keys participants)
+        (extractNamedActions currentSlot executionState)
   in
     state { previousSteps = previousSteps, namedActions = namedActions }
 
+-- This helper function expands actions that can be taken by anybody,
+-- then groups by participant and sorts it so that the owner starts first and the rest go
+-- in alphabetical order
+expandAndGroupByRole ::
+  Set Party ->
+  Set Party ->
+  Array NamedAction ->
+  Array (Tuple Party (Array NamedAction))
+expandAndGroupByRole userParties allParticipants actions =
+  expandedActions
+    # Array.sortBy currentPartiesFirst
+    # Array.groupBy sameParty
+    # map extractGroupedParty
+  where
+  -- If an action has a participant, just use that, if it doesn't expand it to all
+  -- participants
+  expandedActions :: Array (Tuple Party NamedAction)
+  expandedActions =
+    actions
+      # foldMap \action -> case getActionParticipant action of
+          Just participant -> [ participant /\ action ]
+          Nothing -> Set.toUnfoldable allParticipants <#> \participant -> participant /\ action
+
+  currentPartiesFirst (Tuple party1 _) (Tuple party2 _)
+    | Set.member party1 userParties = LT
+    | Set.member party2 userParties = GT
+    | otherwise = compare party1 party2
```

Comment:
Comparison functions should be antisymmetric. In this case it is possible that:
`currentPartiesFirst a b = LT`
but also
`currentPartiesFirst b a = LT`
So that is a bit odd. We could do this instead, for example:
```haskell
  isUserParty party = Set.member party userParties

  currentPartiesFirst (Tuple party1 _) (Tuple party2 _)
    | isUserParty party1 == isUserParty party2 = compare party1 party2
    | otherwise = if isUserParty party1 then LT else GT
```

Reply:
Good catch 👍🏻